File: Modules/Todo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.IO;
using System.Timers;
using SteamKit2;

namespace SteamNerd.Modules
{
    class Todo : Module
    {
        private const double SAVE_TIME = 60000;
        private static readonly string PATH = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            @"SteamNerd"
        );
        private const string FILE_NAME = @"todo.txt";

        private List<string> _todoList;
        private bool _changed;
        private Timer _saveTimer;

        public Todo(SteamNerd steamNerd) : base(steamNerd)
        {
            Name = "TODO";
            Description = "Things to do.";

            AddCommand(
                "todo",
                "Print the todo list.",
                GetTodo
            );

            // Dummy command
            AddCommand(
                "todo add [thing to do]",
                "Adds a thing to do to the todo list.",
                null
            );

            // Dummy command
            AddCommand(
                "todo add [position] [thing to do]",
                "Adds a thing to do to the todo list at a position.",
                null
            );

            // Dummy command
            AddCommand(
                "todo remove [position]",
                "Adds a thing to do to the todo list.",
                null
            );

            _todoList = new List<string>();

            _changed = false;
            _saveTimer = new Timer(SAVE_TIME);
            _saveTimer.Elapsed += (src, e) => Save();
            _saveTimer.Start();

            Load();
        }

        /// <summary>
        /// Save the todo list
        /// </summary>
        private void Save()
        {
            if (_changed)
            {
                var path = Path.Combine(PATH, FILE_NAME);

                using (var fileStream = File.Open(path, FileMode.Create))
                {
                    var writer = new BinaryWriter(fileStream);
                    writer.Write(_todoList.Count);

                    foreach (var todo in _todoList)
                    {
                        writer.Write(todo);
                    }

                    writer.Flush();
                }

                _changed = false;
            }
        }

        /// <summary>
        /// Load the todo list
        /// </summary>
        private void Load()
        {
            var path = Path.Combine(PATH, FILE_NAME);

            if (!Directory.Exists(PATH))
            {
                Directory.CreateDirectory(PATH);
            }

            if (!File.Exists(path))
            {
                File.Create(path);
                return;
            }

            using (var fileStream = File.Open(path, FileMode.Open))
            {
                var reader = new BinaryReader(fileStream);
                uint count = 0;

                try
                {
                    count = reader.ReadUInt32();
                }
                // File is empty
                catch (EndOfStreamException)
                {
                    return;
                }

                for (var i = 0; i < count; i++)
                {
                    _todoList.Add(reader.ReadString());
                }
            }
        }

        public void GetTodo(SteamFriends.ChatMsgCallback callback, string[] args)
        {
            if (args.Length == 1)
            {
                PrintTodo(callback);
            }
            else
            {
                var subcommand = args[1];

                switch (subcommand)
                {
                    case "add":
                        AddTodo(callback.ChatRoomID, args);
                        break;
                    case "remove":
                        RemoveTodo(callback.ChatRoomID, args);
                        break;
                    default:
                        var message = string.Format("Unknown subcommand. Use {0}help for help.", SteamNerd.CommandChar);
                        SteamNerd.SendMessage(message, callback.ChatRoomID, true);
                        break;
                }
            }
        }

        private void PrintTodo(SteamFriends.ChatMsgCallback callback)
        {
            var chat = callback.ChatRoomID;

            if (_todoList.Count == 0)
            {
                SteamNerd.SendMessage("Wow! There's nothing to do!", chat, true);
            }
            else
            {
                var message = "TODO:\n";

                for (var i = 0; i < _todoList.Count; i++)
                {
                    // The most real i
                    var theRealI = i + 1;

                    message += string.Format("{0}. {1}\n", theRealI, _todoList[i]);
                }

                SteamNerd.SendMessage(message, chat, true);
            }
        }

        private void AddTodo(SteamID chat, string[] args)
        {
            if (args.Length < 3)
            {
                return;
            }

            int position;
            if (int.TryParse(args[2], out position))
            {
                if (position > _todoList.Count || position < 1)
                {
                    position = -1;
                }
            }
            else
            {
                position = -1;
            }

            var temp = new List<string>();
            var sentence = "";
            var skip = position != -1 ? 3 : 2;

            foreach (var word in args.Skip(skip))
            {
                if (word.Contains("\n"))
                {
                    sentence += word.Replace("\n", "");
                    temp.Add(sentence.Trim());

                    sentence = "";
                }
                else
                {
                    sentence += word + " ";
                }
            }

            temp.Add(sentence.Trim());

            if (position == -1)
            {
                _todoList.AddRange(temp);
            }
            else
            {
                _todoList.InsertRange(position - 1, temp);
            }

            _changed = true;
        }

        private void RemoveTodo(SteamID chat, string[] args)
        {
            if (args.Length < 2)
            {
                SteamNerd.SendMessage("Invalid position!", chat, true);
            }

            int position;
            if (!int.TryParse(args[2], out position) || position > _todoList.Count || position < 1)
            {
                SteamNerd.SendMessage("Invalid position!", chat, true);
                return;
            }

            _todoList.RemoveAt(position - 1);
            _changed = true;
        }
    }
}

File: Modules/Todo.py
import os
import cPickle as pickle
from threading import Timer

Module.Name = "TODO"
Module.Description = "Things to do."
Module.Global = True

var.TodoList = []
var.File = "TODO.p"
var.Path = os.path.join(os.getenv('APPDATA'), "SteamNerd", var.File)
var.Changed = False

def LoadTodo():
	try:
		var.TodoList = pickle.load(open(var.Path, 'rb'))
	except:
		pass

def SaveTodo():
	if var.Changed:
		var.Changed = False
		pickle.dump(var.TodoList, open(var.Path, 'wb'))

def PrintTodo(callback, args):
	message = ""
	
	if len(var.TodoList) == 0:
		message = "Wow! There's nothing to do!"
	else:
		message = "TODO:\n"
		
		for i in xrange(len(var.TodoList)):
			message += "{}. {}\n".format(i + 1, var.TodoList[i])
		
	Say(message, callback.ChatRoomID)

def AddTodo(callback, args):
	if len(args) < 3:
		Say(
			"Usage: {0}todo add [item] or {0}todo add [line number] [item]".format(SteamNerd.CommandChar),
			callback.ChatRoomID
		)
		return

	line = CheckNumber(args[2])
	
	tempList = []
	tempItem = ""
	skip = 2 if line < 0 else 3
	
	for arg in args[skip:]:
		if arg.find('\n') < 0:
			tempItem += arg + " "
		else:
			tempList.append(tempItem + arg.replace('\n', ''))
			tempItem = ""
	
	tempList.append(tempItem)
	
	if line >= 0:
		var.TodoList = var.TodoList[:line] + tempList + var.TodoList[line:]
	else:
		var.TodoList.extend(tempList)
		
	var.Changed = True

def RemoveTodo(callback, args):
	if len(args) < 3:
		Say(
			"Usage: {}todo remove [line number]".format(SteamNerd.CommandChar),
			callback.ChatRoomID
		)
		return
		
	line = CheckNumber(args[2])
	
	if line >= 0:
		del var.TodoList[line]
		var.Changed = True

def CheckNumber(arg):
	line = 0
		
	try:
		line = int(arg) - 1
	except ValueError:
		return -1
	
	if line < 0 or line > len(var.TodoList):
		return -1
	
	return line

LoadTodo()
saveTimer = Timer(60, SaveTodo)
saveTimer.start()
	
Module.AddCommand("todo", "Print the todo list.", PrintTodo)

Module.AddCommand(
	("todo", "add"), 
	"Add to the todo list. Usage: {}todo add [item]".format(SteamNerd.CommandChar),
	AddTodo
)
Module.AddCommand(
	("todo", "remove"),
	"Removes a thing to do from the todo list. Usage: {}todo remove [item number]".format(SteamNerd.CommandChar),
	RemoveTodo
)
